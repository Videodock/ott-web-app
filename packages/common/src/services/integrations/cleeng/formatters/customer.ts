import type { CleengCustomer } from '../types/models';
import type { Customer } from '../../../../../types/account';

export const formatCustomer = (customer: CleengCustomer): Customer => {
  return {
    id: customer.id,
    email: customer.email,
    country: customer.country,
    firstName: customer.firstName,
    lastName: customer.lastName,
    fullName: `${customer.firstName} ${customer.lastName}`,
    // map `externalData` to `metadata` (NOTE; The Cleeng API returns parsed values)
    metadata: customer.externalData || {},
  };
};
