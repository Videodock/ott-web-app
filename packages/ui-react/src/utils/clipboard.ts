export const copyToClipboard = (value: string): void => {
  const inputElement = document.createElement('input');
  inputElement.style.zIndex = '-10';
  inputElement.style.position = 'absolute';
  inputElement.style.left = '-1000';
  inputElement.value = value;
  document.body.appendChild(inputElement);
  inputElement.select();
  document.execCommand('copy');
  inputElement.blur();
  document.body.removeChild(inputElement);
};
