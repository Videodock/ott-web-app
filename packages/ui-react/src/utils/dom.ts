export const canUseDOM = !!(typeof window !== 'undefined' && window.document && window.document.createElement);

let size: number;

export default function scrollbarSize(recalc?: boolean) {
  if ((!size && size !== 0) || recalc) {
    if (canUseDOM) {
      const scrollDiv = document.createElement('div');
      scrollDiv.style.position = 'absolute';
      scrollDiv.style.top = '-9999px';
      scrollDiv.style.width = '50px';
      scrollDiv.style.height = '50px';
      scrollDiv.style.overflow = 'scroll';
      document.body.appendChild(scrollDiv);
      size = scrollDiv.offsetWidth - scrollDiv.clientWidth;
      document.body.removeChild(scrollDiv);
    }
  }
  return size;
}

export const addScript = (src: string): Promise<void> => {
  return new Promise((resolve) => {
    const script: HTMLScriptElement = document.createElement('script');
    script.type = 'text/javascript';
    script.async = true;
    script.src = src;
    script.onload = () => resolve();
    script.onerror = (error) => {
      console.info('Error loading external script', error);
      throw new Error('Error loading external script');
    };
    document.body.appendChild(script);
  });
};
