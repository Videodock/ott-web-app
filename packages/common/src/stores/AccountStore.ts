import type { CustomFormField, Customer, ConsentsValue } from '../../types/account';
import type { Offer } from '../../types/checkout';
import type { PaymentDetail, Subscription, Transaction } from '../../types/subscription';

import { createStore } from './utils';

type AccountStore = {
  loading: boolean;
  user: Customer | null;
  subscription: Subscription | null;
  transactions: Transaction[] | null;
  activePayment: PaymentDetail | null;
  consents: CustomFormField[] | null;
  consentsValues: ConsentsValue[] | null;
  registrationFields: CustomFormField[] | null;
  registrationFieldsValues: Record<string, string | boolean> | null;
  // True when registration fields are enabled
  registrationFieldsEnabled: boolean;
  // True when the registration fields need to be shown on the create account step
  registrationFieldsOnSignUp: boolean;
  pendingOffer: Offer | null;
  setLoading: (loading: boolean) => void;
  getAccountInfo: () => { customerId: string; customer: Customer; customerConsents: ConsentsValue[] | null };
};

export const useAccountStore = createStore<AccountStore>('AccountStore', (set, get) => ({
  loading: true,
  user: null,
  subscription: null,
  transactions: null,
  activePayment: null,
  consentsValues: null,
  consents: null,
  registrationFields: null,
  registrationFieldsEnabled: false,
  registrationFieldsOnSignUp: false,
  registrationFieldsValues: null,
  pendingOffer: null,
  setLoading: (loading: boolean) => set({ loading }),
  getAccountInfo: () => {
    const user = get().user;
    const customerConsents = get().consentsValues;

    if (!user?.id) throw new Error('user not logged in');

    return { customerId: user?.id, customer: user, customerConsents };
  },
}));
