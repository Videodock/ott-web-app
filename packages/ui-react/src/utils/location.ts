import { createURL, type QueryParamsArg } from '@jwp/ott-common/src/utils/urlFormatting';
import type { Location } from 'react-router-dom';

import type { AccountModals } from '../containers/AccountModal/AccountModal';

export const createURLFromLocation = (location: Location, queryParams: QueryParamsArg = {}) => {
  return createURL(`${location.pathname}${location.search || ''}`, queryParams);
};

export const modalURLFromLocation = (location: Location, u: keyof AccountModals | null, queryParams?: QueryParamsArg) => {
  return createURL(`${location.pathname}${location.search || ''}`, { u, ...queryParams });
};
