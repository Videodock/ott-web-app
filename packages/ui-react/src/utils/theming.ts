import type { Config } from '@jwp/ott-common/types/config';
import { calculateContrastColor } from '@jwp/ott-common/src/utils/common';
import env from '@jwp/ott-common/src/env';

export const setThemingVariables = (config: Config) => {
  const root = document.querySelector(':root') as HTMLElement;
  const { backgroundColor, highlightColor, headerBackground } = config.styling || {};
  const bodyFont = env.APP_BODY_FONT;
  const bodyAltFont = env.APP_BODY_ALT_FONT;

  if (!root) return;

  if (backgroundColor) {
    root.style.setProperty('--body-background-color', backgroundColor);
    root.style.setProperty('--background-contrast-color', calculateContrastColor(backgroundColor));
  }
  if (highlightColor) {
    root.style.setProperty('--highlight-color', highlightColor);
    root.style.setProperty('--highlight-contrast-color', calculateContrastColor(highlightColor));
  }
  if (headerBackground) {
    root.style.setProperty('--header-background', headerBackground);
    root.style.setProperty('--header-contrast-color', calculateContrastColor(headerBackground));
  }

  if (bodyFont) {
    root.style.setProperty('--body-font-family', bodyFont);
  }
  if (bodyAltFont) {
    root.style.setProperty('--body-alt-font-family', bodyAltFont);
  }
};
