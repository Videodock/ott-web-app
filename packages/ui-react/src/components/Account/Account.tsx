import React, { useMemo, useState } from 'react';
import { Trans, useTranslation } from 'react-i18next';
import { useLocation, useNavigate } from 'react-router-dom';
import { shallow } from '@jwp/ott-common/src/utils/compare';
import DOMPurify from 'dompurify';
import { useMutation, useQuery } from 'react-query';
import { getModule } from '@jwp/ott-common/src/modules/container';
import { useAccountStore } from '@jwp/ott-common/src/stores/AccountStore';
import { isTruthy, logDev } from '@jwp/ott-common/src/utils/common';
import AccountController from '@jwp/ott-common/src/controllers/AccountController';
import useToggle from '@jwp/ott-hooks-react/src/useToggle';
import Visibility from '@jwp/ott-theme/assets/icons/visibility.svg?react';
import VisibilityOff from '@jwp/ott-theme/assets/icons/visibility_off.svg?react';
import { formatConsentsValues } from '@jwp/ott-common/src/utils/collection';
import env from '@jwp/ott-common/src/env';

import type { FormSectionContentArgs, FormSectionProps } from '../Form/FormSection';
import Alert from '../Alert/Alert';
import Button from '../Button/Button';
import Form from '../Form/Form';
import IconButton from '../IconButton/IconButton';
import TextField from '../TextField/TextField';
import Checkbox from '../Checkbox/Checkbox';
import HelperText from '../HelperText/HelperText';
import CustomRegisterField from '../CustomRegisterField/CustomRegisterField';
import Icon from '../Icon/Icon';
import { modalURLFromLocation } from '../../utils/location';
import { useAriaAnnouncer } from '../../containers/AnnouncementProvider/AnnoucementProvider';

import styles from './Account.module.scss';

type Props = {
  panelClassName?: string;
  panelHeaderClassName?: string;
  canUpdateEmail?: boolean;
};

const Account = ({ panelClassName, panelHeaderClassName, canUpdateEmail = true }: Props): JSX.Element => {
  const accountController = getModule(AccountController);

  const { t, i18n } = useTranslation('user');
  const announce = useAriaAnnouncer();
  const navigate = useNavigate();
  const location = useLocation();
  const [viewPassword, toggleViewPassword] = useToggle();
  const [isAlertVisible, setIsAlertVisible] = useState(false);
  const exportData = useMutation(accountController.exportAccountData, {
    onSettled: () => setIsAlertVisible(true),
  });
  const exportDataMessage = exportData.isSuccess ? t('account.export_data_success') : t('account.export_data_error');
  const htmlLang = i18n.language !== env.APP_DEFAULT_LANGUAGE ? env.APP_DEFAULT_LANGUAGE : undefined;

  useQuery(['consents'], accountController.getConsents);
  useQuery(['consentsValues'], accountController.getConsentsValues);
  useQuery(['registrationFields'], accountController.getRegistrationFields);

  const { customer, consentsFields, consentsFormValues, registrationFields, registrationFieldsValues } = useAccountStore(
    (s) => ({
      customer: s.user,
      consentsFields: s.consents || [],
      consentsFormValues: formatConsentsValues(s.consentsValues || []),
      registrationFields: s.registrationFields || [],
      registrationFieldsValues: s.registrationFieldsValues || {},
    }),
    shallow,
  );

  const { canChangePasswordWithOldPassword, canExportAccountData, canDeleteAccount } = accountController.getFeatures();
  // users authenticated with social (register_source: Facebook, Google, X (Twitter)) do not have password by default
  const registerSource = customer?.metadata?.register_source;
  const isSocialLogin = (registerSource && registerSource !== 'inplayer') || false;
  const shouldAddPassword = (isSocialLogin && !customer?.metadata?.has_password) || false;

  const initialValues = useMemo(
    () => ({
      consents: consentsFormValues,
      registrationFields: registrationFieldsValues,
      updateEmail: {
        confirmationPassword: '',
        email: customer?.email || '',
      },
    }),
    [consentsFormValues, registrationFieldsValues, customer?.email],
  );

  const formatConsentLabel = (label: string): string | JSX.Element => {
    const sanitizedLabel = DOMPurify.sanitize(label);
    const hasHrefOpenTag = /<a(.|\n)*?>/.test(sanitizedLabel);
    const hasHrefCloseTag = /<\/a(.|\n)*?>/.test(sanitizedLabel);

    if (hasHrefOpenTag && hasHrefCloseTag) {
      return <span dangerouslySetInnerHTML={{ __html: label }} />;
    }

    return label;
  };

  function translateErrors(errors?: string[]): Record<string, string> {
    const formErrors: Record<string, string> = { form: '' };
    // Some errors are combined in a single CSV string instead of one string per error
    errors
      ?.flatMap((e) => e.split(','))
      .forEach((error) => {
        switch (error.trim()) {
          case 'Invalid param email':
            formErrors.email = t('account.errors.invalid_param_email');
            break;
          case 'Customer email already exists':
            formErrors.email = t('account.errors.email_exists');
            break;
          case 'Please enter a valid e-mail address.':
            formErrors.email = t('account.errors.please_enter_valid_email');
            break;
          case 'Invalid confirmationPassword': {
            formErrors.confirmationPassword = t('account.errors.invalid_password');
            break;
          }
          case 'firstName can have max 50 characters.': {
            formErrors.firstName = t('account.errors.first_name_too_long');
            break;
          }
          case 'lastName can have max 50 characters.': {
            formErrors.lastName = t('account.errors.last_name_too_long');
            break;
          }
          case 'Email update not supported': {
            formErrors.form = t('account.errors.email_update_not_supported');
            break;
          }
          default: {
            formErrors.form = t('account.errors.unknown_error');
            logDev('Unknown error', error);
            break;
          }
        }
      });

    return formErrors;
  }

  function formSection(props: FormSectionProps<typeof initialValues, Record<string, string>>) {
    return {
      ...props,
      className: panelClassName,
      panelHeaderClassName: panelHeaderClassName,
      saveButton: t('account.save'),
      cancelButton: t('account.cancel'),
      content: (args: FormSectionContentArgs<typeof initialValues, string[]>) => {
        // This function just allows the sections below to use the FormError type instead of an array of errors
        // @TODO move this (translate errors) to controller
        const formErrors = translateErrors(args.errors);

        // Render the section content, but also add a warning text if there's a form level error
        return (
          <>
            {props.content?.({ ...args, errors: formErrors })}
            <HelperText error={!!formErrors?.form}>{formErrors?.form}</HelperText>
          </>
        );
      },
    };
  }

  const editPasswordClickHandler = async () => {
    if (!customer) {
      return;
    }
    if (isSocialLogin && shouldAddPassword) {
      await accountController.resetPassword(customer.email, '');
      return navigate(modalURLFromLocation(location, 'add-password'));
    }

    navigate(modalURLFromLocation(location, canChangePasswordWithOldPassword ? 'edit-password' : 'reset-password'));
  };

  return (
    <>
      <h1 className={styles.hideUntilFocus}>{t('nav.account')}</h1>

      <Form initialValues={initialValues}>
        {[
          formSection({
            label: t('account.about_you'),
            editButton: t('account.edit_information'),
            onSubmit: async (values) => {
              if (!registrationFields) {
                throw new Error('Registration fields are not loaded');
              }
              await accountController.updateRegistrationFieldsValues(registrationFields, values.registrationFields);

              announce(t('account.update_success', { section: t('account.about_you') }), 'success');
            },
            content: (section) => (
              <>
                {registrationFields?.map((field) => (
                  <CustomRegisterField
                    key={field.name}
                    type={field.type}
                    name={`registrationFields.${field.name}`}
                    value={section.values.registrationFields[field.name] || ''}
                    onChange={section.onChange}
                    placeholder={field.placeholder}
                    disabled={!section.isEditing || section.isBusy}
                    label={field.label}
                    required={field.required}
                    options={field.options}
                    lang={htmlLang}
                  />
                ))}
              </>
            ),
          }),
          formSection({
            label: t('account.email'),
            onSubmit: async (values) => {
              await accountController.updateUser({
                email: values.updateEmail.email || '',
                confirmationPassword: values.updateEmail.confirmationPassword,
              });

              announce(t('account.update_success', { section: t('account.email') }), 'success');
            },
            canSave: (values) => !!(values.updateEmail.email && values.updateEmail.confirmationPassword),
            editButton: t('account.edit_account'),
            readOnly: !canUpdateEmail,
            content: (section) => (
              <>
                <TextField
                  name="updateEmail.email"
                  label={t('account.email')}
                  value={section.values.updateEmail?.email || ''}
                  onChange={section.onChange}
                  // @TODO after moving error handling to controller
                  // error={!!section.errors?.email}
                  // helperText={section.errors?.email}
                  disabled={section.isBusy}
                  editing={section.isEditing}
                  required
                />
                {section.isEditing && (
                  <TextField
                    name="updateEmail.confirmationPassword"
                    label={t('account.confirm_password')}
                    value={section.values.updateEmail.confirmationPassword}
                    onChange={section.onChange}
                    error={!!section.errors?.confirmationPassword}
                    helperText={section.errors?.confirmationPassword}
                    type={viewPassword ? 'text' : 'password'}
                    disabled={section.isBusy}
                    rightControl={
                      <IconButton aria-label={viewPassword ? t('account.hide_password') : t('account.view_password')} onClick={() => toggleViewPassword()}>
                        <Icon icon={viewPassword ? Visibility : VisibilityOff} />
                      </IconButton>
                    }
                    required
                  />
                )}
              </>
            ),
          }),
          formSection({
            label: t('account.security'),
            editButton: (
              <Button
                label={shouldAddPassword ? t('account.add_password') : t('account.edit_password')}
                type="button"
                onClick={() => (customer ? editPasswordClickHandler() : null)}
              />
            ),
          }),
          formSection({
            label: t('account.terms_and_tracking'),
            saveButton: t('account.update_consents'),
            onSubmit: async (values) => {
              await accountController.updateConsents(values.consents);

              announce(t('account.update_success', { section: t('account.terms_and_tracking') }), 'success');
            },
            content: (section) => (
              <>
                {consentsFields?.map((consent, index) => (
                  <Checkbox
                    key={index}
                    name={`consents.${consent.name}`}
                    checked={section.values.consents?.[consent.name]}
                    onChange={section.onChange}
                    label={formatConsentLabel(consent.label)}
                    disabled={consent.required || section.isBusy}
                    lang={htmlLang}
                  />
                ))}
              </>
            ),
          }),
          canExportAccountData &&
            formSection({
              label: t('account.export_data_title'),
              content: () => (
                <div className={styles.textWithButtonContainer}>
                  <div>
                    <Trans t={t} i18nKey="account.export_data_body" values={{ email: customer?.email || '' }} />
                  </div>
                  <div>
                    <Button
                      label={t('account.export_data_title')}
                      type="button"
                      disabled={exportData.isLoading}
                      onClick={async () => {
                        exportData.mutate();
                      }}
                    />
                  </div>
                </div>
              ),
            }),
          canDeleteAccount &&
            formSection({
              label: t('account.delete_account.title'),
              content: () => (
                <div className={styles.textWithButtonContainer}>
                  <div>{t('account.delete_account.body')}</div>
                  <div>
                    <Button
                      label={t('account.delete_account.title')}
                      type="button"
                      variant="danger"
                      onClick={() => {
                        navigate(modalURLFromLocation(location, shouldAddPassword ? 'warning-account-deletion' : 'delete-account'));
                      }}
                    />
                  </div>
                </div>
              ),
            }),
        ].filter(isTruthy)}
      </Form>
      {canExportAccountData && (
        <Alert open={isAlertVisible} message={exportDataMessage} onClose={() => setIsAlertVisible(false)} isSuccess={exportData.isSuccess} />
      )}
    </>
  );
};

export default Account;
