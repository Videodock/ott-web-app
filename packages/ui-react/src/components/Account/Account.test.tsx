import React from 'react';
import type { CustomFormField } from '@jwp/ott-common/types/account';
import { useAccountStore } from '@jwp/ott-common/src/stores/AccountStore';
import AccountController from '@jwp/ott-common/src/stores/AccountController';
import customer from '@jwp/ott-testing/fixtures/customer.json';
import { mockService } from '@jwp/ott-common/test/mockService';
import { DEFAULT_FEATURES } from '@jwp/ott-common/src/constants';

import { renderWithRouter } from '../../../test/utils';

import Account from './Account';

describe('<Account>', () => {
  beforeEach(() => {
    // TODO: remove AccountController from component
    mockService(AccountController, { getFeatures: () => DEFAULT_FEATURES });
  });

  test('renders and matches snapshot', () => {
    useAccountStore.setState({
      user: customer,
      publisherConsents: Array.of({ name: 'marketing', label: 'Receive Marketing Emails' } as CustomFormField),
    });

    const { container } = renderWithRouter(<Account panelClassName={'panel-class'} panelHeaderClassName={'header-class'} canUpdateEmail={true} />);

    // todo
    expect(container).toMatchSnapshot();
  });
});
