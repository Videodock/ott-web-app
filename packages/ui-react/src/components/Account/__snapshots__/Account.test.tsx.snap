// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Account> > renders and matches snapshot 1`] = `
<div>
  <h1
    class="hideUntilFocus"
  >
    nav.account
  </h1>
  <section
    aria-labelledby="account_about_you_1235"
    class="panel-class"
  >
    <div
      class="header-class"
    >
      <h2
        id="account_about_you_1235"
      >
        account.about_you
      </h2>
    </div>
    <form
      class="_flexBox_1c1c63"
      novalidate=""
    >
      <h1
        class="_hideUntilFocus_c968f7"
      >
        nav.account
      </h1>
      <div
        class="_textField_e16c1b"
      >
        <label
          class="_label_e16c1b"
          for="text-field_1235_firstname"
        >
          account.firstname
        </label>
        <input
          aria-describedby="helper_text_1235_firstname"
          aria-invalid="false"
          aria-required="false"
          class="_input_e16c1b"
          id="text-field_1235_firstname"
          lang="en"
          name="firstName"
          readonly=""
          type="text"
          value="John"
        />
      </div>
      <div
        class="_textField_e16c1b"
      >
        <label
          class="_label_e16c1b"
          for="text-field_1235_lastname"
        >
          account.lastname
        </label>
        <input
          aria-describedby="helper_text_1235_lastname"
          aria-invalid="false"
          aria-required="false"
          class="_input_e16c1b"
          id="text-field_1235_lastname"
          lang="en"
          name="lastName"
          readonly=""
          type="text"
          value="Doe"
        />
      </div>
    </form>
    <div
      class="_controls_1c1c63"
    >
      <button
        class="_button_f8f296 _default_f8f296 _outlined_f8f296"
        type="button"
      >
        <span>
          account.edit_information
        </span>
      </button>
    </div>
  </section>
  <section
    aria-labelledby="account_email_1235"
    class="panel-class"
  >
    <div
      class="header-class"
    >
      <h2
        id="account_email_1235"
      >
        account.email
      </h2>
    </div>
    <form
      class="_flexBox_1c1c63"
      novalidate=""
    >
      <div
        class="_textField_e16c1b"
      >
        <label
          class="_label_e16c1b"
          for="text-field_1235_email"
        >
          account.email
        </label>
        <input
          aria-describedby="helper_text_1235_email"
          aria-invalid="false"
          aria-required="true"
          class="_input_e16c1b"
          id="text-field_1235_email"
          name="email"
          readonly=""
          required=""
          type="text"
          value="email@domain.com"
        />
      </div>
    </form>
    <div
      class="_controls_1c1c63"
    >
      <button
        class="_button_f8f296 _default_f8f296 _outlined_f8f296"
        type="button"
      >
        <span>
          account.edit_account
        </span>
      </button>
    </div>
  </section>
  <section
    aria-labelledby="account_security_1235"
    class="panel-class"
  >
    <div
      class="header-class"
    >
      <h2
        id="account_security_1235"
      >
        account.security
      </h2>
    </div>
    <form
      class="_flexBox_1c1c63"
      novalidate=""
    />
    <div
      class="_controls_1c1c63"
    >
      <button
        class="_button_f8f296 _default_f8f296 _outlined_f8f296"
        type="button"
      >
        <span>
          account.edit_password
        </span>
      </button>
    </div>
  </section>
  <section
    aria-labelledby="account_terms_and_tracking_1235"
    class="panel-class"
  >
    <div
      class="header-class"
    >
      <h2
        id="account_terms_and_tracking_1235"
      >
        account.terms_and_tracking
      </h2>
    </div>
    <form
      class="_flexBox_1c1c63"
      novalidate=""
    >
      <div
        class="_checkbox_531f07"
      >
        <div
          class="_row_531f07"
        >
          <input
            aria-describedby="helper_text_1235_consentsvalues_marketing"
            id="check-box_1235_consentsvalues_marketing"
            name="consentsValues.marketing"
            type="checkbox"
            value=""
          />
          <label
            for="check-box_1235_consentsvalues_marketing"
            lang="en"
          >
            Receive Marketing Emails
          </label>
        </div>
      </div>
    </form>
    <div
      class="_controls_1c1c63"
    />
  </section>
</div>
`;
